src/Dofs/DirichletBoundaryConditions2.jl

```diff
@@ -0,0 +1,342 @@
+A function is also given that should be of the form `(x,t) -> v` where `x` is the coordinate of the node, `t` is a
+time parameter and `v` should be of the same length as the number of components the bc is applied to:
+
+```jldoctest
+julia> addnodeset!(grid, "clamped", x -> norm(x[1]) ≈ 0.0);
+
+julia> nodes = grid.nodesets["clamped"]
+
+julia> push!(dbc, :temperature, nodes, (x,t) -> t * [x[2], 0.0, 0.0], [1, 2, 3])
+```
+
+Boundary conditions are now updates by specifying a time:
+
+```jldoctest
+julia> t = 1.0;
+
+julia> update!(dbc, t)
+```
+
+The boundary conditions can be applied to a vector:
+
+```jldoctest
+julia> u = zeros(ndofs(dh))
+
+julia> apply!(u, dbc)
+```
+
+"""
+struct DirichletBoundaryCondition # <: Constraint
+    f::Function # f(x,t) -> value
+    faces::Set{Tuple{Int,Int}}
+    field_name::Symbol
+    components::Vector{Int} # components of the field
+    local_face_dofs::Vector{Int}
+    local_face_dofs_offset::Vector{Int}
+end
+function DirichletBoundaryCondition(field_name::Symbol, faces::Set{Tuple{Int,Int}}, f::Function, component::Int=1)
+    DirichletBoundaryCondition(field_name, faces, f, [component])
+end
+function DirichletBoundaryCondition(field_name::Symbol, faces::Set{Tuple{Int,Int}}, f::Function, components::Vector{Int})
+    unique(components) == components || error("components not unique: $components")
+    # issorted(components) || error("components not sorted: $components")
+    return DirichletBoundaryCondition(f, faces, field_name, components, Int[], Int[])
+end
+
+struct DirichletBoundaryConditions{DH<:DofHandler,T}
+    dbcs::Vector{DirichletBoundaryCondition}
+    prescribed_dofs::Vector{Int}
+    # prescribed_dofs_offset::Vector{Int}
+    free_dofs::Vector{Int}
+    values::Vector{T}
+    dofmapping::Dict{Int,Int} # global dof -> index into dofs and values
+    dh::DH
+    closed::ScalarWrapper{Bool}
+end
+
+function DirichletBoundaryConditions(dh::DofHandler)
+    @assert isclosed(dh)
+    DirichletBoundaryConditions(DirichletBoundaryCondition[], Int[], #=Int[],=# Int[], Float64[], Dict{Int,Int}(), dh, ScalarWrapper(false))
+end
+
+function Base.show(io::IO, dbcs::DirichletBoundaryConditions)
+    println(io, "DirichletBoundaryConditions:")
+    if !isclosed(dbcs)
+        print(io, "  Not closed!")
+    else
+        println(io, "  BCs:")
+        for dbc in dbcs.dbcs
+            println(io, "    ", "Field: ", dbc.field_name, ", ", "Components: ", dbc.components)
+        end
+    end
+end
+
+isclosed(dbcs::DirichletBoundaryConditions) = dbcs.closed[]
+free_dofs(dbcs::DirichletBoundaryConditions) = dbcs.free_dofs
+prescribed_dofs(dbcs::DirichletBoundaryConditions) = dbcs.prescribed_dofs
+
+function close!(dbcs::DirichletBoundaryConditions)
+    fdofs = setdiff(1:ndofs(dbcs.dh), dbcs.prescribed_dofs)
+    copy!!(dbcs.free_dofs, fdofs)
+    # TODO: I *think* it is allowed to sort!(dbcs.prescribed_dofs) here,
+    # then we should be able to utilize that in apply! and friends
```

Comment:
Any reason this would not work?

Reply:
It should work, although the gain is probably dwarfed by the amount of work that's done in `apply!`

Reply:
Right, but since we sort there anyway we might as well just do it here instead.